LwM2M SenML CBOR content format

VeijoPesonen opened the pull request:
This PR builds on top of the PR #44943 which brought in plain CBOR support. 

* Adds the content format.
* Replaces SenML JSON as the default content format due to a smaller data size

Commit:
net: lwm2m: SenML CBOR RW support

* LwM2M v1.1 SenML CBOR content format support.
* Composite operations
* SEND functionality

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

Commit:
net: lwm2m: SenML CBOR gencode adaption

Generated code needs some modifications for it to work correctly.

Adds license information.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

Commit:
net: lwm2m: enable SenML CBOR

With LwM2M v1.1 SenML CBOR is preferred over SenML JSON.

TLV is on by default only with v1.0.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

Commit:
net: lwm2m: empty CBOR array support

Makes possible to write an empty CBOR array if there are no SenML CBOR
records that needs to be written. This came up when trying to delete a
portfolio object instance.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

Commit:
net: lwm2m: Sets ct to registration msg

Tells the server which content format is preferred in a registration
message.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

VeijoPesonen commented:
Fixed a bug found by one of the PR checks.

VeijoPesonen commented:
@rlubos, @SeppoTakalo, @oyvindronningstad, @jheiskan81, @jarlamsa Please review.

rlubos commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -442,6 +442,20 @@ config LWM2M_RW_CBOR_SUPPORT
+	  Include support for writing SenML CBOR data
+
+config LWM2M_RW_SENML_CBOR_RECORDS
+	int "Maximum # of SenML records packed into a CBOR binary" if LWM2M_RW_SENML_CBOR_SUPPORT
```
Please use `depends on LWM2M_RW_SENML_CBOR_SUPPORT` instead of if.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_senml_cbor_decode.c:
```diff
@@ -0,0 +1,207 @@
+/*
+ * Generated using zcbor version 0.3.99
+ * https://github.com/NordicSemiconductor/zcbor
+ * Generated with a --default-max-qty of 99
```
Shouldn't this mention `https://github.com/zephyrproject-rtos/zcbor` instead? The same goes for other generated files.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+	uint_fast8_t dret;
+	int len;
+	int ret;
+
```
Extra emtpy line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+	char objlnk[sizeof("65535:65535")] = {0};
+	unsigned long id;
+	int ret;
+
```
Extra empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+		opaque->remaining = fd->current->_record_union._union_vd.len;
+
+		fd->current = NULL;
+		goto NOT_SUPPORTED;
```
Please use lowercase for labels.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+			break;
+		}
+	}
+
```
Extra empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+	};
+};
+
+struct cbor_in_fmt_data {
```
Would be good to make the naming consistent - either use `formatter` or `fmt` in both cases, `cbor_in_` and `cbor_out_`.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,901 @@
+
+static struct cbor_out_formatter_data *setup_out_fmt_data(struct lwm2m_message *msg)
+{
+	struct cbor_out_formatter_data *fd = malloc(sizeof(*fd));
```
Do we really need to involve heap here? Couldn't senml cbor do the same as json, i. e. allocate the context on a stack?

VeijoPesonen replied:
I tried with it and ended up needing 14K of stack space... If someone finds a way to reduce the memory consumption then using stack might be an option. I don't have any ideas about how to do it at the moment though.

rlubos replied:
What's so large, `struct lwm2m_senml`? Anyway, maybe a better option then would be to have a single preallocated in/out context and protect it with a mutex? I think we could even keep the contexts in a union to save space. The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps). 

I'd really would like to prevent a situation where the SenML cbor encoder is rendered useless due to some other component hogging heap in the system.

VeijoPesonen replied:
`struct lwm2m_senml`, yeah, that's the one.

VeijoPesonen replied:
@oyvindronningstad Any good ideas on how to optimize the code?

VeijoPesonen replied:
> The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps).
Would it make sense if a SEND operation would be executed by using the same queue?

rlubos replied:
I'm not sure I follow, the encoding in `lwm2m_engine_send()` takes place in the caller thread, I don't think there's an easy way to delegate this operation to LwM2M thread. Once encoded, the LwM2M message is queued and processed by the LwM2M thread (i. e. sent over a socket).

rlubos replied:
And just to clarify - I don't think that it's a big deal that encoding can take place from a different thread. If we implment a proper mutex protection for the shared senml cbor context, the only effect would be a slight delay in one of the threads.

oyvindronningstad replied:
> Any good ideas on how to optimize the code?

From my rough count, `struct lwm2m_senml` should take about 3k of space (30 * ~100). You can likely save a few bytes on each record by not using `numeric`:

Change
```
value = tstr / bstr / numeric / bool
```
to
```
value = tstr / bstr / int .size 8 / float / bool
```

You can also try changing
```
	? ( v => numeric  //   ; Numeric Value
```
to
```
	? ( v => int .size 8  //   ; Numeric Value
		v => float //   ; Numeric Value
```
Though that might increase code size.

VeijoPesonen replied:
> You can also try changing
> 
> ```
> 	? ( v => numeric  //   ; Numeric Value
> ```
> 
> to
> 
> ```
> 	? ( v => int .size 8  //   ; Numeric Value
> 		v => float //   ; Numeric Value
> ```

That doesn't seem to work
```
zcbor --default-max-qty 99 -c senml_stripped_v2.cddl code -d -t lwm2m_senml --oc lwm2m_senml_cbor_decode2.c --oh lwm2m_senml_cbor_decode2.h
Parsing files: senml_stripped_v2.cddl, /home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/cddl/prelude.cddl
Traceback (most recent call last):
  File "/home/vep1/.virtualenvs/zephyr/bin/zcbor", line 8, in <module>
    sys.exit(main())
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2960, in main
    args.process(args)
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2884, in process_code
    renderer = CodeRenderer(entry_types={mode: [cddl_res[mode].my_types[entry]
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2388, in __init__
    self.functions[mode] = self.unique_funcs(mode)
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2433, in unique_funcs
    assert func_names[func_name][0] == func_xcode,\
AssertionError: Two elements share the function name decode_union_v, but their implementations are not identical. Please change one or both names.

(((zcbor_uint32_expect(state, (2))))
	&& (zcbor_int64_decode(state, (&(*result)._union_v)))
	&& ((((*result)._union_v >= -9223372036854775807LL)
	&& ((*result)._union_v <= 9223372036854775807LL)) || (zcbor_error(state, ZCBOR_ERR_WRONG_RANGE), false)))

(((zcbor_uint32_expect(state, (2))))
	&& (zcbor_float_decode(state, (&(*result)._union_v))))
[ncs2/zephyr|zephyr|]:~/ws$ cat senml_stripped_v2.cddl 
lwm2m_senml = [1* record]

record = {
	? bn => tstr,           ; Base Name
	? n => tstr,            ; Name
	? ( v => int .size 8 // ; Numeric Value
	    v => float       // ; Numeric Value
	    vs => tstr       // ; String Value
	    vb => bool       // ; Boolean Value
	    vd => bstr ),       ; Data Value
	0*3 key-value-pair      ; To handle unordered maps; length-first ordered map keys
}

; now define the generic versions
key-value-pair = ( int => value )

value = tstr / bstr / int .size 8 / float / bool

n  = 0
bn = -2
v  = 2
vs = 3
vd = 8
vb = 4
````

oyvindronningstad replied:
This seems to work (using two different aliases for `v`):

```
lwm2m_senml = [1* record]

record = {
	? bn => tstr,           ; Base Name
	? n => tstr,            ; Name
	? ( vi => int .size 8 // ; Numeric Value
	    vf => float       // ; Numeric Value
	    vs => tstr       // ; String Value
	    vb => bool       // ; Boolean Value
	    vd => bstr ),       ; Data Value
	0*3 key-value-pair      ; To handle unordered maps; length-first ordered map keys
}

; now define the generic versions
key-value-pair = ( int => value )

value = tstr / bstr / int .size 8 / float / bool

n  = 0
bn = -2
vi  = 2
vf  = 2
vs = 3
vd = 8
vb = 4
```

VeijoPesonen replied:
Another thing that I have been pondering is that instead of having int64s should I turn those into int32s and uint32s. The mandatory objects won't need 64-bit values but do we need to be prepared that someone would introduce an object which would use those?

VeijoPesonen replied:
@oyvindronningstad I regenerated the code based on your suggestions and made the corresponding adjustments.

oyvindronningstad replied:
>instead of having int64s should I turn those into int32s and uint32s

@VeijoPesonen This would not save you any space, since they are in unions with tstr/bstr/float which use 8 bytes.

VeijoPesonen replied:
> What's so large, `struct lwm2m_senml`? Anyway, maybe a better option then would be to have a single preallocated in/out context and protect it with a mutex? I think we could even keep the contexts in a union to save space. The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps).

@rlubos Should be done now as you requested.

Commit:
net: lwm2m: adds README

Includes instructions for generating the SenML CBOR encoder and
decoder.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

Commit:
net: lwm2m: SenML CBOR optimisations

Regenerates the encoder and decoder. Treats integers and floating-point
values as separate entities instead of saying that those are numerical
values. Brings some memory savings.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

oyvindronningstad commented on subsys/net/lib/lwm2m/README_lwm2m:
```diff
@@ -0,0 +1,50 @@
+### SenML CBOR
+
+SenML CBOR content format support takes advantage of the Zcbor library's capability of generating
+domain specific code. To generate the SenML encoder and decoder the following CDDL file
```
I imagine a future you or me will be glad if you made these instructions into a bash script instead :).

oyvindronningstad commented on subsys/net/lib/lwm2m/README_lwm2m:
````diff
@@ -0,0 +1,50 @@
+zcbor --default-max-qty 99 -c CDDL_FILE code -d -t lwm2m_senml --oc lwm2m_senml_cbor_decode.c --oh lwm2m_senml_cbor_decode.h
+```
+
+The default value of the number of maximum allowed records doesn't really matter as it's replaced
+with a Kconfig option.
````
Please describe any steps taken so they can be correctly done the next time this is regenerated.

VeijoPesonen replied:
That's why I put those steps into a separate commit - https://github.com/zephyrproject-rtos/zephyr/pull/44945/commits/1c7e08ef786eda281ac88ce63486e4e53977763e . I'll try to accommodate the same info somehow to this README file.

VeijoPesonen replied:
It would be neat if it would be possible to include either a string or an integer when defining max quantity. I could have fed the Kconfig option directly in the code generation phase.

```
#define DEFAULT_MAX_QTY CONFIG_LWM2M_RW_SENML_CBOR_RECORDS
```

oyvindronningstad replied:
https://github.com/NordicSemiconductor/zcbor/pull/216

VeijoPesonen replied:
High level description given.

oyvindronningstad replied:
Sorry, I cannot find the description.

Commit:
net: lwm2m: use stack for SenML CBOR data

Formatter data are is shared between different threads.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

oyvindronningstad commented on subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c:
```diff
@@ -0,0 +1,893 @@
+/*
+ * SEND is called from a different context than the rest of the LwM2M functionality
+ */
+K_MUTEX_DEFINE(fd_mtx);
```
Curious, why not have separate mutexes for `i` and `o`?

oyvindronningstad replied:
Disregard, I didn't notice that `cbor_io_fmt_data` was a union.

Commit:
net: lwm2m: SenML CBOR key ordering

Keys' order - bn, n - might differ from the default due to
Length-First Map Key Ordering rules.

Signed-off-by: Veijo Pesonen <veijo.pesonen@nordicsemi.no>

VeijoPesonen commented:
> @rlubos It seems that I have introduced a bug on composite write somehow. I'll inform you once I have fixed it.

It was like a missing feature than a bug. The keys might get reordered based on the Length-First Map Key Ordering rules.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_senml_cbor_types.h:
```diff
@@ -18,6 +18,16 @@
 #include <string.h>
 #include "zcbor_common.h"
 
+enum lwm2m_senml_cbor_key {
```
@oyvindronningstad It would be nice if the generator would create these automatically.
